### PR TITLE
feat: add multi-factor authentication (MFA) feature

### PR DESCRIPTION
**Pull request by leo220yuyaodog**
Fix: https://github.com/casdoor/casdoor/issues/81

- [x] Two-factor authentication interface
- [x] Two-factor authentication API
- [x] Add accountItem 
- [x]  Two-facto ui in user page 
- [x]  Setup two-factor UI
- [x] Setup two-factor backend 
- [x] Complete the authenticate process
- [x] Improve code qulity and error handle
- [x] i18n

### CODE REVIEW

**Comment by leo220yuyaodog** [comment]
Now, the basic function has complete. Here is a video. Next, I will to test all the scenario to make sure the program doesn't crash.

@hsluoyz 
By the way, what do you think could be improved about the UI? At present, the UI is only a prototype that can basically complete the function, I will continue to optimize.

https://user-images.githubusercontent.com/71440988/235618952-8593faf4-687a-4616-aee3-476fa30a6573.mp4

File: authz/authz.go [hunk 1]
```diff
@@ -123,6 +123,7 @@ p, *, *, GET, /api/get-release, *, *
 p, *, *, GET, /api/get-default-application, *, *
 p, *, *, GET, /api/get-prometheus-info, *, *
 p, *, *, *, /api/metrics, *, *
+p, *, *, POST, /api/two-factor/auth/verity, *, *
```
Comment (hsluoyz) [inline authz/authz.go:126]: /api/2fa/verify

File: controllers/base.go [hunk 2]
```diff
@@ -168,6 +168,32 @@ func (c *ApiController) SetSessionData(s *SessionData) {
 	c.SetSession("SessionData", util.StructToJson(s))
 }
 
+func (c *ApiController) setTfaSessionData(data *object.TwoFactorSessionData) {
```
Comment (hsluoyz) [inline controllers/base.go:171]: Tfa -> 2fa for all

File: controllers/two_factor.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
+package controllers
```
Comment (hsluoyz) [inline controllers/two_factor.go:1]: Apache header

File: object/tfa.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package object
```
Comment (hsluoyz) [inline object/tfa.go:1]: Apache

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

File: controllers/two_factor.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+)
+
+// TwoFactorSetupInitiate
+// @Title TwoFactorSetupInitiate
+// @Tag Two-Factor API
+// @Description setup totp
+// @param userId	form	string	true	" "<owner>/<name>" of user"
+// @Success 200 {object}   The Response object
+// @router /mfa/setup/initiate [post]
+func (c *ApiController) TwoFactorSetupInitiate() {
```
Comment (hsluoyz) [inline controllers/two_factor.go:32]: MfaSetupInitiate

File: controllers/two_factor.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+// TwoFactorSetupVerify
+// @Title TwoFactorSetupVerify
+// @Tag Two-Factor API
+// @Description setup verify totp
+// @param	secret		form	string	true	"totp secret"
+// @param	passcode	form 	string 	true	"totp passcode"
+// @Success 200 {object}  Response object
+// @router /mfa/setup/totp/verify [post]
+func (c *ApiController) TwoFactorSetupVerify() {
```
Comment (hsluoyz) [inline controllers/two_factor.go:74]: MfaSetupVerify

File: object/mfa.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+import (
+	"fmt"
+
+	"github.com/casdoor/casdoor/util"
+
+	"github.com/beego/beego/context"
+)
+
+type TwoFactorSessionData struct {
```
Comment (hsluoyz) [inline object/mfa.go:25]: MfaSessionData

File: object/mfa.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	Id            string   `json:"id,omitempty"`
+	IsPreferred   bool     `json:"isPreferred,omitempty"`
+	AuthType      string   `json:"type,omitempty" form:"type"`
+	Secret        string   `json:"secret,omitempty"`
+	CountryCode   string   `json:"countryCode,omitempty"`
+	URL           string   `json:"url,omitempty"`
+	RecoveryCodes []string `json:"recoveryCodes,omitempty"`
+}
+
+type TwoFactorInterface interface {
```
Comment (hsluoyz) [inline object/mfa.go:39]: All `TwoFactor` to `Mfa`

File: object/user.go [hunk 9]
```diff
@@ -733,3 +740,35 @@ func (user *User) refreshAvatar() bool {
 
 	return false
 }
+
+func (user *User) IsEnableTwoFactor() bool {
```
Comment (hsluoyz) [inline object/user.go:744]: IsMfaEnabled

File: routers/router.go [hunk 10]
```diff
@@ -238,6 +238,11 @@ func initAPI() {
 	beego.Router("/api/webauthn/signin/begin", &controllers.ApiController{}, "Get:WebAuthnSigninBegin")
 	beego.Router("/api/webauthn/signin/finish", &controllers.ApiController{}, "Post:WebAuthnSigninFinish")
 
+	beego.Router("/api/mfa/setup/initiate", &controllers.ApiController{}, "POST:TwoFactorSetupInitiate")
+	beego.Router("/api/mfa/setup/verify", &controllers.ApiController{}, "Post:TwoFactorSetupVerify")
+	beego.Router("/api/mfa/setup/enable", &controllers.ApiController{}, "POST:TwoFactorSetupEnable")
+	beego.Router("/api/mfa", &controllers.ApiController{}, "DELETE:TwoFactorDelete")
```
Comment (hsluoyz) [inline routers/router.go:244]: why not use POST + `/api/delete-mfa`?

File: util/validation.go [hunk 11]
```diff
@@ -67,3 +67,21 @@ func GetCountryCode(prefix string, phone string) (string, error) {
 
 	return countryCode, nil
 }
+
+func GetCountryCodeFromE164Number(phone string) (string, error) {
```
Comment (hsluoyz) [inline util/validation.go:71]: Is `phone` with prefix? Merge with previous `GetCountryCode()`

File: web/.eslintrc [hunk 12]
```diff
@@ -82,6 +82,12 @@
       }
     ],
     "no-multi-spaces": ["error", { "ignoreEOLComments": true }],
+    "react/no-unknown-property": [
```
Comment (hsluoyz) [inline web/.eslintrc:85]: Is this for debug use? Don't commit it then

File: web/src/App.js [hunk 13]
```diff
@@ -560,6 +561,7 @@ class App extends Component {
         <Route exact path="/payments/:paymentName" render={(props) => this.renderLoginIfNotLoggedIn(<PaymentEditPage account={this.state.account} {...props} />)} />
         <Route exact path="/payments/:paymentName/result" render={(props) => this.renderLoginIfNotLoggedIn(<PaymentResultPage account={this.state.account} {...props} />)} />
         <Route exact path="/records" render={(props) => this.renderLoginIfNotLoggedIn(<RecordListPage account={this.state.account} {...props} />)} />
+        <Route exact path="/two_factor_authentication/setup" render={(props) => this.renderLoginIfNotLoggedIn(<MfaSetupPage account={this.state.account} {...props} />)} />
```
Comment (hsluoyz) [inline web/src/App.js:564]: `/mfa/setup`, we don't use underscore in URL path

File: web/src/OrganizationListPage.js [hunk 14]
```diff
@@ -63,6 +63,7 @@ class OrganizationListPage extends BaseListPage {
         {name: "Roles", visible: true, viewRule: "Public", modifyRule: "Immutable"},
         {name: "Permissions", visible: true, viewRule: "Public", modifyRule: "Immutable"},
         {name: "3rd-party logins", visible: true, viewRule: "Self", modifyRule: "Self"},
+        {Name: "Two-factor authentication", Visible: true, ViewRule: "Self", ModifyRule: "Self"},
```
Comment (hsluoyz) [inline web/src/OrganizationListPage.js:66]: Multi-factor authentication

File: web/src/UserEditPage.js [hunk 15]
```diff
@@ -142,6 +147,58 @@ class UserEditPage extends React.Component {
...
+      RemoveTwoFactorLoading: true,
+    });
+
+    twoFactorRemoveTotp({
+      id: id,
+      owner: this.state.user.owner,
+      name: this.state.user.name,
+    }).then((res) => {
+      if (res.status === "ok") {
+        Setting.showMessage("success", i18next.t("mfa:Removed successfully"));
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:188]: Use existing i18n words

File: web/src/UserEditPage.js [hunk 16]
```diff
@@ -142,6 +147,58 @@ class UserEditPage extends React.Component {
...
+        this.setState({
+          twoFactorAuth: res.data,
+        });
+      } else {
+        Setting.showMessage("error", i18next.t("mfa:Removed failed"));
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:193]: Use existing i18n words

File: web/src/UserEditPage.js [hunk 17]
```diff
@@ -695,6 +752,56 @@ class UserEditPage extends React.Component {
...
       );
+    } else if (accountItem.name === "Two-factor authentication") {
+      return (
+        !this.isSelfOrAdmin() ? null : (
+          <Row style={{marginTop: "20px"}} >
+            <Col style={{marginTop: "5px"}} span={Setting.isMobile() ? 22 : 2}>
+              {Setting.getLabel(i18next.t("mfa:Two-factor authentication"), i18next.t("mfa:Two-factor authentication - Tooltip "))} :
+            </Col>
+            <Col span={22} >
+              <Card title="Two-factor methods">
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:763]: All texts should be put into i18n, or don't add them

File: web/src/UserEditPage.js [hunk 18]
```diff
@@ -695,6 +752,56 @@ class UserEditPage extends React.Component {
...
+                      </List.Item>
+                    )}
+                  />
+                </Card>
+                <Card style={{marginTop: 16}}
+                  type="inner"
+                  title="Authenticator app"
+                >
+                  {this.getMfa(TotpMfaType)}
+                        TOTP developing...
```
Comment (hsluoyz) [inline web/src/UserEditPage.js:798]: Don't add placeholder

File: web/src/auth/MfaVerifyForms.js [hunk 19]
```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Casdoor Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline web/src/auth/MfaVerifyForms.js:1]: Form (no s).js

File: web/src/backend/OrganizationBackend.js [hunk 20]
```diff
@@ -35,11 +35,10 @@ export function getOrganization(owner, name) {
 }
 
 export function updateOrganization(owner, name, organization) {
-  const newOrganization = Setting.deepCopy(organization);
```
Comment (hsluoyz) [inline web/src/backend/OrganizationBackend.js:38]: Why delete this?

File: web/src/locales/de/data.json [hunk 21]
```diff
@@ -367,6 +368,38 @@
     "Text": "Text",
     "Text - Tooltip": "Text - Tooltip"
   },
+  "mfa": {
+    "Each time you sign in to your Account, you'll need your password and a authentication code": "Each time you sign in to your Account, you'll need your password and a authentication code",
+    "Enable": "Enable",
```
Comment (hsluoyz) [inline web/src/locales/de/data.json:373]: A lot of words can be found duplicated with existing words. Try to use existing words as much as possible

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: routers/router.go (hunk 10 above)
Comment (hsluoyz) [reply routers/router.go:244]: Will `Post` work?

image

File: web/.eslintrc (hunk 12 above)
Comment (leo220yuyaodog) [reply web/.eslintrc:85]: It is for ignore css property in theme component. After update package.json, yarn build seem not work without   
this eslint config.

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**